cli/command/volume/mount_linux.go

```diff
@@ -86,95 +96,117 @@ func runMount(storageosCli *command.StorageOSCli, opt mountOptions) error {
 	}
 
 	// checking readiness
-	if err := isVolumeReady(vol, name); err != nil {
+	if err := isVolumeReady(vol); err != nil {
 		return fmt.Errorf("cannot mount volume: %v", err)
 	}
 
+	// Create a context which times out the mount
+	ctx, cancel := context.WithTimeout(context.Background(), opt.timeout)
+	defer cancel()
+
 	err = client.VolumeMount(types.VolumeMountOptions{
-		ID: vol.ID, Namespace: namespace,
+		Context: ctx,
+		ID:      vol.ID, Namespace: namespace,
 		Client:     hostname,
 		Mountpoint: opt.mountpoint,
 		FsType:     vol.FSType,
 	})
 	if err != nil {
-		return err
+		select {
+		case <-ctx.Done():
+			// Server should handle what to do if we close the connection whilst waiting for a response
+			return fmt.Errorf("timed out waiting for volume mount lock")
+		default:
+			return err
+		}
 	}
 
 	fst, err := mount.ParseFSType(vol.FSType)
 	if err != nil {
 		return err
 	}
 
-	err = retryableMount(vol, node.DeviceDir, opt.mountpoint, fst)
+	err = retryableMount(ctx, vol, node.DeviceDir, opt, fst)
 	if err != nil {
-		log.WithFields(log.Fields{
-			"namespace":  namespace,
-			"volumeName": name,
-			"error":      err,
-		}).Error("error while mounting volume")
-		// should unmount volume in the CP if we failed here
-		newErr := client.VolumeUnmount(types.VolumeUnmountOptions{ID: vol.ID, Namespace: namespace})
-		if newErr != nil {
+		select {
+		case <-ctx.Done():
+			client.VolumeUnmount(types.VolumeUnmountOptions{ID: vol.ID, Namespace: namespace})
```

Comment:
Is it possible to handle a disconnected client on the API side and automatically do this? Seems weird to push the responsibility onto a client (especially as it just gave up on a long-running request).

Also will be a bit weird if the API is just hanging in general - a timeout is eventually reached and then we block to unmount too.

Might not be possible - just thinking of alternatives!

Reply:
The retry-able mount is something which doesn't communicate with the API, it runs mount commands on the host. Guaranteeing the lock is released eventually if the client fails to perform the mount would be non-trivial to my understanding.

Reply:
Yeah after our chat this seems to be the best solution currently, thanks!